File: sdks/flutter/lib/survey.dart
import 'package:flutter/material.dart';
import 'likerts.dart';
import 'choice_features.dart';
import 'visibility.dart';
import 'branching.dart';
import 'advanced_questions.dart';

/// Host-provided respondent copy for localization.
class LikertsSurveyStrings {
  final String requiredLabel, submitLabel, datePlaceholder, selectAnswer, backLabel, nextLabel,moveUp,moveDown;
  final String Function(int, int, int) progressLabel;
  final String Function(String) answerRequired, otherError;
  final String Function(int) chooseAtLeast, chooseAtMost;
  final String Function(int) remaining;
  final String Function(String) advancedError;

  const LikertsSurveyStrings({
    this.requiredLabel = 'required',
    this.backLabel = 'Back',
    this.nextLabel = 'Next',
    this.progressLabel = _defaultProgress,
    this.selectAnswer = 'Select an answer',
    this.otherError = _defaultOther,
    this.submitLabel = 'Submit',
    this.datePlaceholder = 'YYYY-MM-DD',
    this.answerRequired = _defaultRequired,
    this.chooseAtLeast = _defaultAtLeast,
    this.chooseAtMost = _defaultAtMost,
    this.moveUp='Move up',this.moveDown='Move down',this.remaining=_defaultRemaining,this.advancedError=_defaultAdvanced,
  });

  static String _defaultProgress(int current, int total, int visited) => 'Page $current of $total · $visited visited';
  static String _defaultOther(String label) => '$label: enter valid text for the selected Other option.';
  static String _defaultRequired(String label) => '$label: an answer is required.';
  static String _defaultAtLeast(int count) => 'Choose at least $count options';
  static String _defaultAtMost(int count) => 'Choose at most $count options';
  static String _defaultRemaining(int count)=>'$count remaining';
  static String _defaultAdvanced(String label)=>'$label: complete the answer.';
}

/// Styling hooks that complement the host application's [ThemeData].
class LikertsSurveyStyle {
  final EdgeInsetsGeometry padding;
  final double questionSpacing, controlSpacing;
  final TextStyle? errorTextStyle;
  final ButtonStyle? submitButtonStyle;

  const LikertsSurveyStyle({
    this.padding = const EdgeInsets.all(16),
    this.questionSpacing = 20,
    this.controlSpacing = 8,
    this.errorTextStyle,
    this.submitButtonStyle,
  });
}

/// Native Flutter renderer. The host owns submission, placement and dismissal.
class LikertsSurvey extends StatefulWidget {
  final Collection collection;
  final bool disabled;
  final LikertsSurveyStrings strings;
  final LikertsSurveyStyle style;
  final ValueChanged<Map<String, dynamic>>? onAnswersChange;
  final ValueChanged<Map<String, dynamic>> onSubmit;

  const LikertsSurvey({
    super.key,
    required this.collection,
    required this.onSubmit,
    this.disabled = false,
    this.strings = const LikertsSurveyStrings(),
    this.style = const LikertsSurveyStyle(),
    this.onAnswersChange,
  });

  @override
  State<LikertsSurvey> createState() => _LikertsSurveyState();
}

class _LikertsSurveyState extends State<LikertsSurvey> {
  final Map<String, dynamic> _answers = {};
  String? _validationError;
  late SurveyFlow _flow;

  @override
  void initState() { super.initState(); _flow = SurveyFlow(widget.collection); }

  @override
  void didUpdateWidget(covariant LikertsSurvey oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.collection.id != widget.collection.id || oldWidget.collection.version != widget.collection.version) {
      _answers.clear();
      _flow = SurveyFlow(widget.collection);
      _validationError = null;
      widget.onAnswersChange?.call(const {});
    }
  }

  void _changed(VoidCallback change) {
    setState(() {
      change();
      final previous = _flow.answers;
      final changed = {...previous.keys, ..._answers.keys}.where((id) => previous[id] != _answers[id]).toList();
      for (final id in changed) { _flow.setAnswer(id, _answers[id]); }
      _answers..clear()..addAll(_flow.answers);
      _validationError = null;
    });
    widget.onAnswersChange?.call(Map.unmodifiable(_answers));
  }

  String? _errorFor(Question question) {
    final value=_answers[question.id];
    if(question.type=='ranking'){if(value==null)return question.required?widget.strings.answerRequired(question.label):null;return validRanking(List<String>.from(value),question.options.map((o)=>o['id'] as String).toList())?null:widget.strings.advancedError(question.label);}
    if(question.type=='matrix'){if(value==null)return question.required?widget.strings.answerRequired(question.label):null;final matrix=(value as Map).map((key,value)=>MapEntry(key as String,value is List?List<String>.from(value):[value as String]));return validMatrix(matrix,question.rows.map((i)=>i.id).toList(),question.columns.map((i)=>i['id'] as String).toList(),multiple:question.matrixMode=='multiple',required:question.required)?null:widget.strings.advancedError(question.label);}
    if(question.type=='constant_sum'){if(value==null)return question.required?widget.strings.answerRequired(question.label):null;return validAllocation(Map<String,int>.from(value),question.items.map((i)=>i.id).toList(),question.total??0)?null:widget.strings.advancedError(question.label);}
    final isChoice=question.type=='single_choice'||question.type=='multiple_choice';
    if(question.required && (value==null||value==''||(isChoice&&selectedChoices(value).isEmpty))) return widget.strings.answerRequired(question.label);
    if(!isChoice) return null;
    final error=choiceError(question,value);if(error==null)return null;
    final minimum=(question.minSelections??0)>(question.required?1:0)?question.minSelections!:(question.required?1:0);
    if(error=='range') return selectedChoices(value).length<minimum?widget.strings.chooseAtLeast(minimum):widget.strings.chooseAtMost(question.maxSelections??question.options.length);
    return widget.strings.otherError(question.label);
  }

  bool get _lastPage => pageRoute(widget.collection, _answers).last == _flow.progress.current - 1;

  Iterable<Question> get _currentQuestions {
    final visible = visibleQuestionIds(widget.collection.questions, _answers);
    return widget.collection.questions.where((q) => _flow.page.questionIds.contains(q.id) && visible.contains(q.id));
  }

  void _advance() {
    final reached = routedAnswers(widget.collection, _answers);
    final visible = visibleQuestionIds(widget.collection.questions, reached);
    final reachedIds = pageRoute(widget.collection, reached).expand((i) => widget.collection.pages?[i].questionIds ?? widget.collection.questions.map((q) => q.id)).toSet();
    final questions = _lastPage ? widget.collection.questions.where((q) => reachedIds.contains(q.id) && visible.contains(q.id)) : _currentQuestions;
    String? error;
    for (final question in questions) { error ??= _errorFor(question); }
    setState(() => _validationError = error);
    if (error != null) return;
    if (_lastPage) { widget.onSubmit(Map.unmodifiable(reached)); }
    else { setState(() { _flow.next(); }); }
  }

  void _back() { setState(() { _flow.back(); _validationError = null; }); }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Semantics(
      container: true,
      child: Padding(
        padding: widget.style.padding,
        child: SingleChildScrollView(child:Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
          Semantics(header: true, child: Text(widget.collection.title, key: const ValueKey('likerts.title'), style: theme.textTheme.headlineSmall)),
          SizedBox(height: widget.style.questionSpacing),
          if (widget.collection.pages != null) ...[
            Semantics(liveRegion: true, child: Text(widget.strings.progressLabel(_flow.progress.current, _flow.progress.total, _flow.progress.visited), key: const ValueKey('likerts.progress'))),
            if (_flow.page.title != null) Semantics(header: true, child: Text(_flow.page.title!, key: const ValueKey('likerts.pageTitle'))),
          ],
          for (final question in _currentQuestions) ...[
            _Question(
              key: ValueKey('likerts.question.${question.id}'),
              collectionKey: '${widget.collection.id}:${widget.collection.version}',
              question: question,
              answer: _answers[question.id],
              disabled: widget.disabled,
              strings: widget.strings,
              controlSpacing: widget.style.controlSpacing,
              onChanged: (value) => _changed(() {
                if (value == null || value == '' || ((question.type=='single_choice'||question.type=='multiple_choice') && selectedChoices(value).isEmpty) || (value is List && value.isEmpty)) {
                  _answers.remove(question.id);
                } else {
                  _answers[question.id] = value;
                }
              }),
            ),
            SizedBox(height: widget.style.questionSpacing),
          ],
          if (_validationError != null)
            Semantics(
              container: true,
              liveRegion: true,
              label: _validationError,
              child: Text(
                _validationError!,
                key: const ValueKey('likerts.error'),
                style: widget.style.errorTextStyle ?? theme.textTheme.bodyMedium?.copyWith(color: theme.colorScheme.error),
              ),
            ),
          SizedBox(height: widget.style.controlSpacing),
          if (_flow.progress.visited > 1) TextButton(key: const ValueKey('likerts.back'), onPressed: widget.disabled ? null : _back, child: Text(widget.strings.backLabel)),
          ElevatedButton(
            key: ValueKey(_lastPage ? 'likerts.submit' : 'likerts.next'),
            onPressed: widget.disabled ? null : _advance,
            style: widget.style.submitButtonStyle,
            child: Text(_lastPage ? widget.strings.submitLabel : widget.strings.nextLabel),
          ),
        ])),
      ),
    );
  }
}

class _Question extends StatelessWidget {
  final String collectionKey;
  final Question question;
  final dynamic answer;
  final bool disabled;
  final LikertsSurveyStrings strings;
  final double controlSpacing;
  final ValueChanged<dynamic> onChanged;

  const _Question({
    super.key,
    required this.collectionKey,
    required this.question,
    required this.answer,
    required this.disabled,
    required this.strings,
    required this.controlSpacing,
    required this.onChanged,
  });

  @override
  Widget build(BuildContext context) => Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
    Semantics(
      header: true,
      child: Text('${question.label}${question.required ? ' (${strings.requiredLabel})' : ''}', style: Theme.of(context).textTheme.titleMedium),
    ),
    SizedBox(height: controlSpacing),
    if (question.presentation == 'dropdown')
      DropdownButton<String>(
        key: ValueKey('likerts.dropdown.${question.id}'),
        value: selectedChoices(answer).isEmpty ? null : selectedChoices(answer).first,
        hint: Text(strings.selectAnswer),
        items: question.options.map((option)=>DropdownMenuItem<String>(value:option['id'],child:Text(option['label']))).toList(),
        onChanged: disabled ? null : (value) { if(value!=null)onChanged(toggleChoice(question,answer,value)); },
      )
    else if(question.type=='ranking')
      ...List<String>.from(answer??question.options.map((o)=>o['id'])).asMap().entries.map((entry){final id=entry.value,index=entry.key,option=question.options.firstWhere((o)=>o['id']==id);final order=List<String>.from(answer??question.options.map((o)=>o['id']));return Row(key:ValueKey('likerts.ranking.${question.id}.$id'),children:[Expanded(child:Text('${index+1}. ${option['label']}')),TextButton(onPressed:disabled||index==0?null:()=>onChanged(moveRanking(order,id,-1)),child:Text(strings.moveUp),),TextButton(onPressed:disabled||index==order.length-1?null:()=>onChanged(moveRanking(order,id,1)),child:Text(strings.moveDown))]);})
    else if(question.type=='matrix')
      ...question.rows.map((row)=>Column(crossAxisAlignment:CrossAxisAlignment.start,children:[
        Text(row.label),
        ...question.columns.map((column){
          final raw=(answer as Map?)?[row.id];
          final selected=raw is List?raw.contains(column['id']):raw==column['id'];
          return CheckboxListTile(
            key:ValueKey('likerts.matrix.${question.id}.${row.id}.${column['id']}'),title:Text(column['label']),value:selected,
            onChanged:disabled?null:(_){
              final current=(answer as Map?)?.map((key,value)=>MapEntry(key as String,value is List?List<String>.from(value):[value as String]))??<String,List<String>>{};
              final next=setMatrixChoice(current,row.id,column['id'],multiple:question.matrixMode=='multiple');
              onChanged(next.map((key,value)=>MapEntry(key,question.matrixMode=='single'?value.first:value)));
            });
        })
      ]))
    else if(question.type=='constant_sum') ...[
      ...question.items.map((item)=>TextFormField(key:ValueKey('$collectionKey:${question.id}:${item.id}'),initialValue:(answer as Map?)?[item.id]?.toString()??'',enabled:!disabled,decoration:InputDecoration(labelText:item.label),keyboardType:TextInputType.number,onChanged:(text){final next=Map<String,int>.from(answer as Map? ?? {});final value=int.tryParse(text);if(value==null){next.remove(item.id);}else{next[item.id]=value;}onChanged(next);})),
      Semantics(liveRegion:true,child:Text(strings.remaining(allocationRemaining(question.total??0,Map<String,int>.from(answer as Map? ?? {}))),key:ValueKey('likerts.remaining.${question.id}')))
    ]
    else if (question.type == 'single_choice')
      ...question.options.map((option) => RadioListTile<String>(
        key: ValueKey('likerts.option.${question.id}.${option['id']}'),
        title: Text(option['label']),
        value: option['id'],
        // ignore: deprecated_member_use
        groupValue: selectedChoices(answer).isEmpty ? null : selectedChoices(answer).first,
        // Keep the pre-3.32 API until the minimum supported Flutter release has RadioGroup.
        // ignore: deprecated_member_use
        onChanged: disabled ? null : (value) { if(value!=null)onChanged(toggleChoice(question,answer,value)); },
      ))
    else if (question.type == 'multiple_choice')
      ...question.options.map((option) {
        final selected = selectedChoices(answer);
        final checked = selected.contains(option['id']);
        final enabled = !disabled && (checked || option['exclusive']==true || question.options.any((o)=>o['exclusive']==true && selected.contains(o['id'])) || selected.length < (question.maxSelections ?? 2147483647));
        return CheckboxListTile(
          key: ValueKey('likerts.option.${question.id}.${option['id']}'),
          title: Text(option['label']),
          value: checked,
          onChanged: !enabled ? null : (_) {
            onChanged(toggleChoice(question,answer,option['id']));
          },
        );
      })
    else if (question.scaleValues.isNotEmpty)
      Wrap(spacing: controlSpacing, runSpacing: controlSpacing, children: question.scaleValues.map((value) => ChoiceChip(
        key: ValueKey('likerts.option.${question.id}.$value'),
        label: Text(question.presentation=='stars'?List.filled(value,'★').join():question.scaleLabel(value),semanticsLabel:question.scaleLabel(value)),
        selected: answer == value,
        onSelected: disabled ? null : (_) => onChanged(value),
      )).toList())
    else
      TextFormField(
        key: ValueKey('$collectionKey:${question.id}'),
        initialValue: answer?.toString() ?? '',
        enabled: !disabled,
        maxLength: question.maxLength,
        decoration: InputDecoration(labelText: question.label, hintText: question.type == 'date' ? strings.datePlaceholder : null),
        keyboardType: question.type == 'number' || question.type == 'scale' ? const TextInputType.numberWithOptions(decimal: true, signed: true) : TextInputType.text,
        onChanged: (value) => onChanged(value.isEmpty ? null : question.type == 'number' || question.type == 'scale' ? num.tryParse(value) ?? value : value),
      ),
    for(final option in question.options.where((o)=>o['other']!=null && selectedChoices(answer).contains(o['id'])))
      TextFormField(
        key: ValueKey('$collectionKey:${question.id}:${option['id']}:other'),
        enabled: !disabled,
        initialValue: otherTexts(answer)[option['id']]??'',
        decoration: InputDecoration(labelText:'${question.label}: ${option['label']}'),
        onChanged: (value)=>onChanged(setOtherText(question,answer,option['id'],value)),
      ),
  ]);
}

File: sdks/android/src/main/kotlin/com/likerts/sdk/Survey.kt
package com.likerts.sdk

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.selection.toggleable
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.*
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlinx.serialization.json.*

/** Host-provided copy. Supplying this value keeps all respondent-facing text localizable. */
data class SurveyStrings(
    val requiredLabel: String = "required",
    val submitLabel: String = "Submit",
    val backLabel: String = "Back",
    val nextLabel: String = "Next",
    val progressLabel: (Int, Int, Int) -> String = { current, total, visited -> "Page $current of $total · $visited visited" },
    val selectAnswer: String = "Select an answer",
    val otherError: (String) -> String = { "$it: enter valid text for the selected Other option." },
    val datePlaceholder: String = "YYYY-MM-DD",
    val chooseAtLeast: (Int) -> String = { "Choose at least $it options" },
    val chooseAtMost: (Int) -> String = { "Choose at most $it options" },
    val answerRequired: (String) -> String = { "$it: an answer is required." },
    val moveUp:String="Move up",val moveDown:String="Move down",val remaining:(Int)->String={"$it remaining"},val advancedError:(String)->String={"$it: complete the answer."},
)

/** Layout and validation colors that can be aligned with the host application's theme. */
data class SurveyStyle(
    val contentPadding: Dp = 16.dp,
    val questionSpacing: Dp = 20.dp,
    val controlSpacing: Dp = 8.dp,
    val errorColor: Color? = null,
)

private fun answers(questions: List<Question>, values: Map<String, String>, choices: Map<String, Set<String>>, otherText: Map<String,Map<String,String>> = emptyMap(),advanced:Map<String,JsonElement> = emptyMap()): Map<String, JsonElement> = buildMap {
    questions.forEach { question ->
        advanced[question.id]?.let{put(question.id,it);return@forEach}
        if (question.type == "multiple_choice") {
            choices[question.id]?.takeIf(Set<String>::isNotEmpty)?.let { put(question.id, question.choiceAnswer(it.sorted(),otherText[question.id].orEmpty())) }
        } else values[question.id]?.takeIf(String::isNotEmpty)?.let { value ->
            val answer = if (question.type == "scale" || question.type == "number") value.toDoubleOrNull()?.takeIf(Double::isFinite)?.let(::JsonPrimitive) ?: JsonPrimitive(value) else if(question.type == "single_choice") question.choiceAnswer(listOf(value),otherText[question.id].orEmpty()) else JsonPrimitive(value)
            put(question.id, answer)
        }
    }
}

private fun visibleAnswers(questions:List<Question>,values:Map<String,String>,choices:Map<String,Set<String>>,otherText:Map<String,Map<String,String>>,advanced:Map<String,JsonElement>):Map<String,JsonElement> = com.likerts.sdk.visibleAnswers(questions,answers(questions,values,choices,otherText,advanced))

private fun validationError(question: Question, textValue: String?, selected: Set<String>?, strings: SurveyStrings, otherText:Map<String,String>,advanced:JsonElement?): String? {
    if(question.type in setOf("ranking","matrix","constant_sum")){
        if(advanced==null)return if(question.required)strings.answerRequired(question.label)else null
        val valid=when(question.type){"ranking"->AdvancedQuestions.validRanking(advanced.jsonArray.map{it.jsonPrimitive.content},question.options.map{it.id});"matrix"->{val value=advanced.jsonObject.mapValues{(_,v)->if(v is JsonArray)v.map{it.jsonPrimitive.content}else listOf(v.jsonPrimitive.content)};AdvancedQuestions.validMatrix(value,question.rows.orEmpty().map{it.id},question.columns.orEmpty().map{it.id},question.matrixMode=="multiple",question.required)};else->{val value=advanced.jsonObject.mapValues{it.value.jsonPrimitive.int};AdvancedQuestions.validAllocation(value,question.items.orEmpty().map{it.id},question.total?:0)}}
        return if(valid)null else strings.advancedError(question.label)
    }
    val missing = if (question.type == "multiple_choice") selected.isNullOrEmpty() else textValue.isNullOrEmpty()
    if (question.required && missing) return strings.answerRequired(question.label)
    if(question.type != "single_choice" && question.type != "multiple_choice") return null
    val ids=if(question.type=="single_choice") listOfNotNull(textValue?.takeIf(String::isNotEmpty)) else selected.orEmpty().toList()
    val error=question.choiceError(ids,otherText) ?: return null
    val minimum=maxOf(question.minSelections ?: 0,if(question.required) 1 else 0)
    return if(error=="range") { if(ids.size<minimum) strings.chooseAtLeast(minimum) else strings.chooseAtMost(question.maxSelections ?: question.options.size) } else strings.otherError(question.label)
}

/** Native Compose renderer. Form state is bound to collection ID and immutable version. */
@Composable fun LikertsSurvey(
    collection: Collection,
    disabled: Boolean = false,
    modifier: Modifier = Modifier,
    strings: SurveyStrings = SurveyStrings(),
    style: SurveyStyle = SurveyStyle(),
    onAnswersChange: ((Map<String, JsonElement>) -> Unit)? = null,
    onSubmit: (Map<String, JsonElement>) -> Unit,
) {
    var formError by remember(collection.id, collection.version) { mutableStateOf<String?>(null) }
    val values = remember(collection.id, collection.version) { mutableStateMapOf<String, String>() }
    val choices = remember(collection.id, collection.version) { mutableStateMapOf<String, Set<String>>() }
    val otherText = remember(collection.id,collection.version) { mutableStateMapOf<String,Map<String,String>>() }
    val dropdowns = remember(collection.id,collection.version) { mutableStateMapOf<String,Boolean>() }
    val advanced=remember(collection.id,collection.version){mutableStateMapOf<String,JsonElement>()}
    val flow = remember(collection.id, collection.version) { SurveyFlow(collection.schema) }
    var flowRevision by remember(collection.id, collection.version) { mutableIntStateOf(0) }
    fun changed() {
        formError = null
        collection.schema.questions.forEach { q ->
            val selected = if (q.type == "single_choice") listOfNotNull(values[q.id]) else choices[q.id].orEmpty().toList()
            otherText[q.id]?.let { text -> otherText[q.id] = text.filterKeys { it in selected } }
        }
        val raw = answers(collection.schema.questions, values, choices, otherText,advanced)
        val previous = flow.answers
        (previous.keys + raw.keys).filter { previous[it] != raw[it] }.forEach { flow.setAnswer(it, raw[it]) }
        val retained = flow.answers.keys
        values.keys.removeAll { it !in retained }; choices.keys.removeAll { it !in retained }; otherText.keys.removeAll { it !in retained };advanced.keys.removeAll{it !in retained}
        flowRevision++
        onAnswersChange?.invoke(flow.answers)
    }
    val page = remember(flow, flowRevision) { flow.page }
    val progress = remember(flow, flowRevision) { flow.progress }
    val lastPage = pageRoute(collection.schema, flow.answers).last() == progress.current - 1


    Column(modifier.padding(style.contentPadding).testTag("likerts.survey"), verticalArrangement = Arrangement.spacedBy(style.questionSpacing)) {
        Text(collection.schema.title, style = MaterialTheme.typography.headlineMedium, modifier = Modifier.semantics { heading() }.testTag("likerts.title"))
        val visible=visibleQuestionIds(collection.schema.questions,answers(collection.schema.questions,values,choices,otherText,advanced))
        if (collection.schema.pages != null) {
            Text(strings.progressLabel(progress.current, progress.total, progress.visited), modifier = Modifier.semantics { liveRegion = LiveRegionMode.Polite }.testTag("likerts.progress"))
            page.title?.let { Text(it, modifier = Modifier.semantics { heading() }.testTag("likerts.pageTitle")) }
        }
        val currentQuestions = collection.schema.questions.filter { it.id in visible && it.id in page.questionIds }
        currentQuestions.forEach { question ->
            Column(Modifier.fillMaxWidth().testTag("likerts.question.${question.id}"), verticalArrangement = Arrangement.spacedBy(style.controlSpacing)) {
                Text(question.label + if (question.required) " (${strings.requiredLabel})" else "", style = MaterialTheme.typography.titleMedium, modifier = Modifier.semantics { heading() })
                when (question.type) {
                    "ranking" -> {val order=advanced[question.id]?.jsonArray?.map{it.jsonPrimitive.content}?:question.options.map{it.id};order.forEachIndexed{index,id->val option=question.options.first{it.id==id};Row(verticalAlignment=Alignment.CenterVertically){Text("${index+1}. ${option.label}",Modifier.weight(1f));TextButton(enabled=!disabled&&index>0,onClick={advanced[question.id]=JsonArray(AdvancedQuestions.move(order,id,-1).map(::JsonPrimitive));changed()},modifier=Modifier.semantics{contentDescription="${strings.moveUp}: ${option.label}"}){Text(strings.moveUp)};TextButton(enabled=!disabled&&index<order.lastIndex,onClick={advanced[question.id]=JsonArray(AdvancedQuestions.move(order,id,1).map(::JsonPrimitive));changed()},modifier=Modifier.semantics{contentDescription="${strings.moveDown}: ${option.label}"}){Text(strings.moveDown)}}}}
                    "matrix" -> question.rows.orEmpty().forEach{row->Column{Text(row.label);question.columns.orEmpty().forEach{column->val objectValue=advanced[question.id]?.jsonObject;val selected=if(question.matrixMode=="single")objectValue?.get(row.id)?.jsonPrimitive?.content==column.id else objectValue?.get(row.id)?.jsonArray?.any{it.jsonPrimitive.content==column.id}==true;Row(Modifier.fillMaxWidth().toggleable(selected,!disabled,if(question.matrixMode=="single")Role.RadioButton else Role.Checkbox){val existing=objectValue?.mapValues{(_,v)->if(v is JsonArray)v.map{it.jsonPrimitive.content}else listOf(v.jsonPrimitive.content)}.orEmpty();val changedValue=AdvancedQuestions.setMatrix(existing,row.id,column.id,question.matrixMode=="multiple");advanced[question.id]=JsonObject(changedValue.mapValues{(_,v)->if(question.matrixMode=="single")JsonPrimitive(v.first())else JsonArray(v.map(::JsonPrimitive))});changed()}.testTag("likerts.matrix.${question.id}.${row.id}.${column.id}"),verticalAlignment=Alignment.CenterVertically){if(question.matrixMode=="single")RadioButton(selected,null,enabled=!disabled)else Checkbox(selected,null,enabled=!disabled);Text(column.label)}}}}
                    "constant_sum" -> {question.items.orEmpty().forEach{item->val current=advanced[question.id]?.jsonObject?.get(item.id)?.jsonPrimitive?.content.orEmpty();OutlinedTextField(current,{text->val values=advanced[question.id]?.jsonObject?.toMutableMap()?:mutableMapOf();text.toIntOrNull()?.let{values[item.id]=JsonPrimitive(it)}?:values.remove(item.id);advanced[question.id]=JsonObject(values);changed()},enabled=!disabled,label={Text(item.label)},modifier=Modifier.fillMaxWidth().testTag("likerts.allocation.${question.id}.${item.id}"))};val allocated=advanced[question.id]?.jsonObject?.values?.sumOf{it.jsonPrimitive.intOrNull?:0}?:0;Text(strings.remaining((question.total?:0)-allocated),Modifier.semantics{liveRegion=LiveRegionMode.Polite}.testTag("likerts.remaining.${question.id}"))}
                    "single_choice" -> if(question.presentation=="dropdown") Box {
                        OutlinedButton(enabled=!disabled,onClick={dropdowns[question.id]=true},modifier=Modifier.testTag("likerts.dropdown.${question.id}")) { Text(question.options.find { it.id==values[question.id] }?.label ?: strings.selectAnswer) }
                        DropdownMenu(expanded=dropdowns[question.id]==true,onDismissRequest={dropdowns[question.id]=false}) {
                            question.options.forEach { option -> DropdownMenuItem(text={Text(option.label)},enabled=!disabled,onClick={values[question.id]=option.id;dropdowns[question.id]=false;changed()},modifier=Modifier.testTag("likerts.option.${question.id}.${option.id}")) }
                        }
                    } else Column(Modifier.selectableGroup()) { question.options.forEach { option ->
                        val selected = values[question.id] == option.id
                        Row(Modifier.fillMaxWidth().selectable(selected, !disabled, Role.RadioButton) { values[question.id] = option.id; changed() }.testTag("likerts.option.${question.id}.${option.id}"), verticalAlignment = Alignment.CenterVertically) {
                            RadioButton(selected, onClick = null, enabled = !disabled); Text(option.label)
                        }
                    } }
                    "multiple_choice" -> question.options.forEach { option ->
                        val selected = choices[question.id].orEmpty(); val checked = option.id in selected
                        val enabled = !disabled && (checked || option.exclusive==true || question.options.any { it.exclusive==true && it.id in selected } || selected.size < (question.maxSelections ?: Int.MAX_VALUE))
                        Row(Modifier.fillMaxWidth().toggleable(checked, enabled, Role.Checkbox) { choices[question.id] = question.toggledChoices(selected.toList(),option.id).toSet(); changed() }.testTag("likerts.option.${question.id}.${option.id}"), verticalAlignment = Alignment.CenterVertically) {
                            Checkbox(checked, onCheckedChange = null, enabled = enabled); Text(option.label)
                        }
                    }
                    "scale" -> if (question.scaleValues().isNotEmpty()) Column(Modifier.selectableGroup()) { question.scaleValues().forEach { value ->
                        val selected = values[question.id] == value.toString()
                        Row(Modifier.fillMaxWidth().selectable(selected, !disabled, Role.RadioButton) { values[question.id] = value.toString(); changed() }.testTag("likerts.option.${question.id}.$value"), verticalAlignment = Alignment.CenterVertically) {
                            RadioButton(selected, onClick = null, enabled = !disabled); Text(if(question.presentation=="stars") "★".repeat(value) else question.scaleLabel(value),modifier=Modifier.semantics { contentDescription=question.scaleLabel(value) })
                        }
                    } } else OutlinedTextField(values[question.id].orEmpty(), { values[question.id] = it; changed() }, enabled = !disabled, label = { Text(question.label) }, modifier = Modifier.fillMaxWidth().testTag("likerts.input.${question.id}"))
                    else -> OutlinedTextField(values[question.id].orEmpty(), { values[question.id] = it; changed() }, enabled = !disabled, label = { Text(question.label) }, placeholder = if (question.type == "date") ({ Text(strings.datePlaceholder) }) else null, modifier = Modifier.fillMaxWidth().testTag("likerts.input.${question.id}"))
                }
                question.options.filter { it.other != null && (if(question.type=="single_choice") values[question.id]==it.id else it.id in choices[question.id].orEmpty()) }.forEach { option ->
                    OutlinedTextField(otherText[question.id]?.get(option.id).orEmpty(),{otherText[question.id]=otherText[question.id].orEmpty()+(option.id to it);changed()},enabled=!disabled,label={Text("${question.label}: ${option.label}")},modifier=Modifier.fillMaxWidth().testTag("likerts.other.${question.id}.${option.id}"))
                }
            }
        }
        formError?.let { message -> Text(message, color = style.errorColor ?: MaterialTheme.colorScheme.error, modifier = Modifier.semantics { liveRegion = LiveRegionMode.Assertive; error(message) }.testTag("likerts.error")) }
        Spacer(Modifier.height(style.controlSpacing))
        if (progress.visited > 1) TextButton(enabled = !disabled, onClick = { flow.back(); flowRevision++; formError = null }, modifier = Modifier.testTag("likerts.back")) { Text(strings.backLabel) }
        Button(enabled = !disabled, onClick = {
            val reachedIds = pageRoute(collection.schema, flow.answers).flatMap { collection.schema.pages?.get(it)?.questionIds ?: collection.schema.questions.map { q -> q.id } }.toSet()
            val toValidate = if (lastPage) collection.schema.questions.filter { it.id in visible && it.id in reachedIds } else currentQuestions
            formError = toValidate.firstNotNullOfOrNull { validationError(it, values[it.id], choices[it.id], strings, otherText[it.id].orEmpty(),advanced[it.id]) }
            if (formError == null) { if (lastPage) onSubmit(routedAnswers(collection.schema, flow.answers)) else { flow.next(); flowRevision++ } }
        }, modifier = Modifier.testTag(if (lastPage) "likerts.submit" else "likerts.next")) { Text(if (lastPage) strings.submitLabel else strings.nextLabel) }
    }
}
